audition/__init__.py

```diff
@@ -1 +1,116 @@
-"""Main application"""
+import yaml
+from smart_open import smart_open
+
+from .distance_from_best import DistanceFromBestTable, BestDistancePlotter
+from .thresholding import ModelGroupThresholder
+from .regrets import SelectionRulePicker, SelectionRulePlotter
+from .selection_rule_performance import SelectionRulePerformancePlotter
+from .model_group_performance import ModelGroupPerformancePlotter
+from .selection_rule_grid import make_selection_rule_grid
+
+
+class Auditioner(object):
+    def __init__(
+        self,
+        db_engine,
+        model_group_ids,
+        train_end_times,
+        initial_metric_filters,
+        models_table=None,
+        distance_table=None,
+    ):
+        self.metric_filters = initial_metric_filters
+        self.train_end_times = train_end_times
+
+        models_table = models_table or 'models'
+        distance_table = distance_table or 'best_distance'
+        self.distance_from_best_table = DistanceFromBestTable(
+            db_engine=db_engine,
+            models_table=models_table,
+            distance_table=distance_table
+        )
+        self.best_distance_plotter = BestDistancePlotter(self.distance_from_best_table)
+        self.model_group_thresholder = ModelGroupThresholder(
+            distance_from_best_table=self.distance_from_best_table,
+            train_end_times=train_end_times,
+            initial_model_group_ids=model_group_ids,
+            initial_metric_filters=initial_metric_filters
+        )
+        self.model_group_performance_plotter = ModelGroupPerformancePlotter(self.distance_from_best_table)
+
+        self.selection_rule_picker = SelectionRulePicker(self.distance_from_best_table)
+        self.selection_rule_plotter = SelectionRulePlotter(self.selection_rule_picker)
+        self.selection_rule_performance_plotter = SelectionRulePerformancePlotter(self.selection_rule_picker)
+
+        self.distance_from_best_table.create_and_populate(
+            model_group_ids,
+            self.train_end_times,
+            self.metrics
+        )
+
+    @property
+    def metrics(self):
+        return [
+            {'metric': f['metric'], 'parameter': f['parameter']}
+            for f in self.metric_filters
+        ]
+
+    @property
+    def thresholded_model_group_ids(self):
+        return self.model_group_thresholder.model_group_ids
+
+    @property
+    def selection_rule_model_group_ids(self):
+        model_group_ids = set()
+        for metric_definition in self.metrics:
+            for selection_rule in self.selection_rules:
+                model_group_ids |= set([
```

Comment:
I don't think we necessarily want the set of all model groups ever chosen by the given selection rule, just the model group chosen in the most recent time frame.

Reply:
Ah, I believe then that the results_for_rule is a confusing thing to call here. Since it's the last time, we don't have a need to try and calculate any regrets, do we?

In fact, we don't really need to be looping through the metrics in this loop, because that metric definition is really just for regrets. Just loop through the selection rules, and call `bound_selection_rule.pick` on each.

Of course, 'results_for_rule' does some setup for bound_selection_rule.pick , in terms of creating the localized dataframe that only has the information that the rule can properly use. But that can be refactored out pretty easily.

Does this sound right?

Reply:
Yeah, I think that sounds right generally -- here we basically just want to apply the selection rules forward and ask what model group they would pick on the final test set, even though we don't have regrets to calculate at that point. Not needing the metrics loop and split out the results_for_rule setup makes sense to me, too.